### ISSUE TITLE
Few-shot Image Generation with Elastic Weight Consolidation

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#90]
*NIPS 2020*

## Motivation

- Few Shot Image Synthesis via Effective Adaptation that combats common over-fitting

## Key Idea

- Among pretrained generator weights, some should be preserved for diversity sake while others should be adapted to address domain appearance gap.
- Study generator finetuning in large data setting, and gain insights in terms of average weight update magnitude and **Fisher Information**.
- Approximate Fisher Information with discriminator output Cross Entropy.
- Update generator weights in inverse proportionate manner according to Fisher Information.

## Previous Approach

- BSA : introduce additional batch normalization layer to BigGAN
- MineGAN : Add another mining network and train in two stage manner

### Goal
![image](https://user-images.githubusercontent.com/46308407/128453309-01d39175-d8ea-4dfc-b11f-45610be64368.png)

### Abundant Data Setting Analogy
![image](https://user-images.githubusercontent.com/46308407/128453332-163428ae-629a-43e3-9a00-ba92e079233e.png)

- When pretraining and finetuning on abundant source and target data, last layers tend to change little, and at the same time they have greater importance in terms of discriminator log likelihood (*hence should be preserved in few shot setting as well*)

### Fisher Information Approximation
![image](https://user-images.githubusercontent.com/46308407/128453419-c800f1d1-aa4d-4616-9764-dc321bc28383.png)
- Equivalent to computing the **Binary Cross Entropy loss on discriminator output**.

### Refined Objective
![image](https://user-images.githubusercontent.com/46308407/128453490-0ea415b9-1cda-4fe8-80bb-9e92922824a9.png)

- Add EWC (Elastic Weight Consolidation) regularization term, that proportionately penalizes weight change from the source model (*Important weights should be preserved*)

## Experiments

![image](https://user-images.githubusercontent.com/46308407/128453586-8ce6f1e2-479c-4a3b-ad41-957f3f4fc364.png)

- Without EWC regularization, weights quickly deviate from the source. (*catastrophic forgetting*)
- In **continual learning framework*, we wish to preserve **diversity** information from the source and newly learn **appearance** or **style** information from the scarce target. (We will forget to generate source domain images)
- Without EWC loss, the fine tuned model quickly overfits the few samples in target domain.

### Baseline Comparison
![image](https://user-images.githubusercontent.com/46308407/128453711-ddee8e31-e347-48a4-b5bd-f46d7f7cfa58.png)
![image](https://user-images.githubusercontent.com/46308407/128453720-e3ce8088-fd69-47cd-87ad-0329f07170a4.png)
![image](https://user-images.githubusercontent.com/46308407/128453731-bf7b8843-af4b-40be-862b-9ba108828931.png)

- When the target domain has few samples, it's implausible to compute FID, so resort to AMT user study.

### Ablation & Analysis

![image](https://user-images.githubusercontent.com/46308407/128453793-4006f127-0a4b-4991-ab77-4e4de720852f.png)
![image](https://user-images.githubusercontent.com/46308407/128453895-985e7efb-882d-4df4-b859-192716f2e741.png)

- Big lambda (=greater penalty on weight deviation) preserves diversity but under-fits new domain
- When source and target domain gets dissimilar, the transfer performance drops (as expected).

![image](https://user-images.githubusercontent.com/46308407/128453855-1a05d2ec-9dce-4c15-ae8a-335af3f29220.png)
- One shot is also possible. It displays certain level of variation learned from source domain pretraining.

![image](https://user-images.githubusercontent.com/46308407/128453906-fb3d89e7-0248-44d6-9e54-525526f1db9f.png)
- Adapted generator shows certain correspondence with pretrained source generator when same latent code has been fed.
- Key to diverse synthesis in few shot target domain.
- Inspiration to #87 (Few Shot Image Generation via Cross-domain Correspondence, CVPR'21)

## Comments
- The idea of distinguishing between *weights to be changed* and *weights to be preserved* is inspiring.
- Also, the way the authors got insight from large data regime analogy is impressive.
- Output result is not that great, compared to #87 , but still it was like the first attempt to generate realistic images with 10 samples, so it deserves credit in that aspect.
